### PR TITLE
Db crud

### PR DESCRIPTION
**Pull request by realwebdev**
Checkout in CRUD complete commit or DB CRUD

### CODE REVIEW

File: conf/conf.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package conf
+
+import (
+	"log"
+
```
Comment (Bilaltanveer38) [inline conf/conf.go:5]: space

File: conf/conf.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+	"github.com/joho/godotenv"
+)
+
+func LoadEnvVar() {
+	err := godotenv.Load()
+	if err != nil {
```
Comment (Bilaltanveer38) [inline conf/conf.go:11]: handle it in a single line

File: database/conn.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package database
+
+import (
+	"fmt"
+	"log"
+	"os"
+
```
Comment (Bilaltanveer38) [inline database/conn.go:7]: space

File: conf/conf.go (hunk 1 above)
Comment (zunnorainzaheer) [inline conf/conf.go:5]: space

File: database/conn.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+	dialect := os.Getenv("DIALECT")
+	host := os.Getenv("HOST")
+	dbPort := os.Getenv("DBPORT")
+	user := os.Getenv("USER")
+	dbname := os.Getenv("NAME")
+	dbpassword := os.Getenv("PASSWORD")
+
+	dbURI := fmt.Sprintf("host=%s user=%s dbname=%s sslmode=disable password=%s port=%s", host, user, dbname, dbpassword, dbPort)
+	db, err := gorm.Open(dialect, dbURI)
+
```
Comment (Bilaltanveer38) [inline database/conn.go:23]: no space required here

File: main.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package main
+
+import (
+	"time"
+
```
Comment (Bilaltanveer38) [inline main.go:5]: space

File: database/conn.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/jinzhu/gorm"
+	_ "github.com/jinzhu/gorm/dialects/postgres"
+)
+
+func ConnectDB() *gorm.DB {
+
```
Comment (Bilaltanveer38) [inline database/conn.go:13]: no space required here

File: models/activity.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	Activity_id   uint `gorm:"primaryKey"`
+	Project_id    uint
+	User_id       uint
+	Activity_name string
+	Start_time    time.Time
+	End_time      time.Time
+	Total_time    time.Time
+}
+
+func CreateActivity(act Activity, db *gorm.DB) {
```
Comment (Bilaltanveer38) [inline models/activity.go:20]: change the argument name to activity instead of act

File: models/activity.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	if err := db.Create(act).Error; err != nil {
+		log.Print("Error Occured")
```
Comment (Bilaltanveer38) [inline models/activity.go:22]: should have return in case of an error, the function should not proceed any further

File: models/activity.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+// 	//start time means save time
+// }
+
+// func endActivityUpdate(time.Time) {
+
+// 	//end time
+// 	//return total time of activity
+// }
+
+func DeleteActivity(actname string, db *gorm.DB) {
```
Comment (Bilaltanveer38) [inline models/activity.go:38]: should have deleted using the primary key not the activity name, many activities can have the same name

File: models/activity.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+	//delete row by id or name
+	if err := db.Where("Activity_name LIKE ?", actname).Delete(Project{}).Error; err != nil {
+		log.Print("Activity not exist")
```
Comment (Bilaltanveer38) [inline models/activity.go:41]: return statement

File: models/activity.go [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+// func endActivityUpdate(time.Time) {
+
+// 	//end time
+// 	//return total time of activity
+// }
+
+func DeleteActivity(actname string, db *gorm.DB) {
+	//delete row by id or name
+	if err := db.Where("Activity_name LIKE ?", actname).Delete(Project{}).Error; err != nil {
```
Comment (Bilaltanveer38) [inline models/activity.go:40]: Why deleting project? makes no sense

File: models/activity.go [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+// }
+
+// func endActivityUpdate(time.Time) {
+
+// 	//end time
+// 	//return total time of activity
+// }
+
+func DeleteActivity(actname string, db *gorm.DB) {
+	//delete row by id or name
```
Comment (Bilaltanveer38) [inline models/activity.go:39]: remove the comment

File: models/activity.go [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+	if err := db.Where("Activity_name LIKE ?", actname).Delete(Project{}).Error; err != nil {
+		log.Print("Activity not exist")
+	}
+	log.Print("Activity successfully deleted")
+}
+
+func UpdateActivity(old_name string, name_update string, db *gorm.DB) {
+	// change activity name
```
Comment (Bilaltanveer38) [inline models/activity.go:47]: remove this comment

File: models/activity.go [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+	if err := db.Model(Project{}).Where("Activity_name = ?", old_name).Updates(Activity{Activity_name: name_update}).Error; err != nil {
```
Comment (Bilaltanveer38) [inline models/activity.go:48]: not the right query to perform the update

File: models/activity.go [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+		log.Print("some error in update")
```
Comment (Bilaltanveer38) [inline models/activity.go:49]: return statement

File: models/project.go [hunk 16]
```diff
@@ -0,0 +1,36 @@
+package models
+
+import (
+	"log"
+
```
Comment (Bilaltanveer38) [inline models/project.go:5]: space

File: models/project.go [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/jinzhu/gorm"
+)
+
+type Project struct {
+	Activity []Activity `gorm:"foreignKey:Project"`
+
```
Comment (Bilaltanveer38) [inline models/project.go:11]: space

File: models/project.go [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+	Project_id   uint `gorm:"primaryKey"`
+	User_id      uint
+	Project_name string
+}
+
+func CreateProject(proj Project, db *gorm.DB) {
+	if err := db.Create(proj).Error; err != nil {
+		log.Print("Error Occured")
```
Comment (Bilaltanveer38) [inline models/project.go:19]: return

File: models/project.go [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+	}
+	log.Print("your project has been created")
+}
+
+func UpdateProject(old_name string, name_update string, db *gorm.DB) {
```
Comment (Bilaltanveer38) [inline models/project.go:24]: should have use the project_id instead of project_old_name

File: models/project.go [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+	if err := db.Model(Project{}).Where("Project_name = ?", old_name).Updates(Project{Project_name: name_update}).Error; err != nil {
+		log.Print("some error in update")
```
Comment (Bilaltanveer38) [inline models/project.go:26]: return

File: models/project.go [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+	}
+	log.Print("successfully updated the name of project")
+}
+
+func DeleteProject(projname string, db *gorm.DB) {
```
Comment (Bilaltanveer38) [inline models/project.go:31]: again use the project_id

File: models/project.go [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+	if err := db.Where("Project_name LIKE ?", projname).Delete(User{}).Error; err != nil {
+		log.Print("Project not exist")
```
Comment (Bilaltanveer38) [inline models/project.go:33]: return

File: models/project.go [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+
+func UpdateProject(old_name string, name_update string, db *gorm.DB) {
+	if err := db.Model(Project{}).Where("Project_name = ?", old_name).Updates(Project{Project_name: name_update}).Error; err != nil {
+		log.Print("some error in update")
+	}
+	log.Print("successfully updated the name of project")
+}
+
+func DeleteProject(projname string, db *gorm.DB) {
+	if err := db.Where("Project_name LIKE ?", projname).Delete(User{}).Error; err != nil {
```
Comment (Bilaltanveer38) [inline models/project.go:32]: syntax is not correct at all

File: models/user.go [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+	Email    string `gorm:"typevarchar(100);unique_index"`
+	Password string
+}
+
+func CreateSchema(db *gorm.DB) {
+	db.AutoMigrate(User{})
+	db.AutoMigrate(Project{})
+	db.AutoMigrate(Activity{})
+}
+
```
Comment (Bilaltanveer38) [inline models/user.go:24]: CreateSchema Function is not according to what i have told you 
it should be like this
AutoMigrate(modelObject interface{}) error{
	if err:=Db.AutoMigrate(modelObject);err!=nil{
		return err
	}

	return nil
}

and it should have been placed in the database package

File: models/user.go [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+func SignUp(user1 User, db *gorm.DB) {
+	if err := db.Create(user1).Error; err != nil {
+		log.Print("wrong username or password")
```
Comment (Bilaltanveer38) [inline models/user.go:27]: Error should be: Error occurred while creating the user in the DB

return statement is missing

File: models/user.go [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+	Password string
+}
+
+func CreateSchema(db *gorm.DB) {
+	db.AutoMigrate(User{})
+	db.AutoMigrate(Project{})
+	db.AutoMigrate(Activity{})
+}
+
+func SignUp(user1 User, db *gorm.DB) {
```
Comment (Bilaltanveer38) [inline models/user.go:25]: user instead of user1

File: models/user.go [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+	if err := db.Create(user1).Error; err != nil {
+		log.Print("wrong username or password")
+	}
+	log.Print("your account has been created")
+}
+
+func SignIn(email string, pass string, db *gorm.DB) {
+	var user User
```
Comment (Bilaltanveer38) [inline models/user.go:33]: use short declaration operator

File: models/user.go [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+func SignUp(user1 User, db *gorm.DB) {
+	if err := db.Create(user1).Error; err != nil {
+		log.Print("wrong username or password")
+	}
+	log.Print("your account has been created")
+}
+
+func SignIn(email string, pass string, db *gorm.DB) {
```
Comment (Bilaltanveer38) [inline models/user.go:32]: SignIn(email, pass string, db *gorm.DB) - signature should be like this

File: models/user.go [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+	var user User
+	if err := db.Where(&User{Email: email, Password: pass}).First(&user).Error; err != nil {
+		log.Print("this user not exist")
```
Comment (Bilaltanveer38) [inline models/user.go:35]: return

File: models/user.go [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+	}
+	log.Print("SgnedIn")
+}
+
+func UserDeletion(email string, db *gorm.DB) {
```
Comment (Bilaltanveer38) [inline models/user.go:40]: use primary key instead of email

File: models/user.go [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+	if err := db.Where("Email LIKE ?", email).Delete(User{}).Error; err != nil {
```
Comment (Bilaltanveer38) [inline models/user.go:41]: not the correct syntax

File: models/user.go [hunk 32]
```diff
@@ -0,0 +1,45 @@
...
+		log.Print("account not exist")
```
Comment (Bilaltanveer38) [inline models/user.go:42]: return

File: models/activity.go (hunk 11 above)
Comment (Bilaltanveer38) [reply models/activity.go:40]: db.Table("activity").Where("id=?", activity_id).Delete(Activity{}).Error

File: models/project.go (hunk 19 above)
Comment (Bilaltanveer38) [reply models/project.go:24]: db.Table("").Where("").Updates(updates).Error

File: models/user.go (hunk 27 above)
Comment (Bilaltanveer38) [reply models/user.go:33]: user:=models.User{} || user:=New(models.User)

File: conf/conf.go (hunk 1 above)
Comment (realwebdev) [reply conf/conf.go:5]: it automatically create space on save the file.

File: conf/conf.go (hunk 2 above)
Comment (realwebdev) [reply conf/conf.go:11]: DOne

File: database/conn.go (hunk 3 above)
Comment (realwebdev) [reply database/conn.go:7]: by saving it creates space

File: database/conn.go (hunk 4 above)
Comment (realwebdev) [reply database/conn.go:23]: done

File: main.go (hunk 5 above)
Comment (realwebdev) [reply main.go:5]: space by default

File: database/conn.go (hunk 6 above)
Comment (realwebdev) [reply database/conn.go:13]: Done

File: models/activity.go (hunk 7 above)
Comment (realwebdev) [reply models/activity.go:20]: Done

File: models/activity.go (hunk 8 above)
Comment (realwebdev) [reply models/activity.go:22]: done

File: models/user.go (hunk 25 above)
Comment (realwebdev) [reply models/user.go:27]: done

File: models/user.go (hunk 26 above)
Comment (realwebdev) [reply models/user.go:25]: done

File: models/user.go (hunk 27 above)
Comment (realwebdev) [reply models/user.go:33]: done

File: models/user.go (hunk 32 above)
Comment (realwebdev) [reply models/user.go:42]: done

File: models/user.go (hunk 29 above)
Comment (realwebdev) [reply models/user.go:35]: done

File: models/user.go (hunk 28 above)
Comment (realwebdev) [reply models/user.go:32]: done

File: models/activity.go (hunk 10 above)
Comment (realwebdev) [reply models/activity.go:41]: DONE

File: models/activity.go (hunk 11 above)
Comment (realwebdev) [reply models/activity.go:40]: delete func was copied forget to change the name to activities.
Done

File: models/activity.go (hunk 9 above)
Comment (realwebdev) [reply models/activity.go:38]: done

**Pull request merged**